manipulation_controller:
  simulation: false
  debug: false
  type: manipulation_royalpanda/ManipulationController
  references_file: $(find mppi_manipulation_royalpanda)/config/experiments/references.yaml
  default_pose: [0.56916, 0.83279, 1.1555, -0.53309, 0.82183, -0.13716, -0.14697]
  object_tolerance: 0.09 # ~5 deg

  policy_update_rate: 0.0  # execute as fast as possible
  reference_update_rate: 10.0
  publish_ros: true
  ros_publish_rate: 10.0
  static_optimization: false

  logging: false
  log_file_path: $(find mppi_manipulation_royalpanda)/logs/$(arg experiment_name)
  fixed_base: false
  arm_id: panda
  joint_names: [ panda_joint1, panda_joint2, panda_joint3, panda_joint4, panda_joint5, panda_joint6, panda_joint7 ]
  base_joint_names: [world_x, world_y, theta]

  state_topic: /observer/state
  nominal_state_topic: /x_nom
  base_twist_topic: /ridgeback_velocity_controller/cmd_vel
  base_cmd_threshold: [0.0, 0.0, 0.0]

  gaussian_policy: true
  sequential: false
  max_position_error: [0.2, 0.2, 0.1, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]

  log_every_steps: 15

  record_bag: false
  bag_path: "/media/giuseppe/My Passport/Work/bags/test.bag"
  
  # Cost parameters used by the controller
  rollouts: 40
  cost:
    regularization: 0.0
    linear_weight: 100.0
    linear_weight_opening: 10.0

    angular_weight: 10.0
    angular_weight_opening: 1.0

    # joint controller cost (mode 3)
    arm_joint_weight: 20
    robot_base_position_weight: 20
    robot_base_orientation_weight: 20

    linear_tolerance_manipulation: 0.0
    angular_tolerance_manipulation: 0.3

    min_dist: 0.15
    max_reach: 0.65
    reach_weight: 100
    reach_weight_slope: 10

    obstacle_weight: 0
    obstacle_weight_slope: 0
    obstacle_radius: 0.8

    contact_weight: 100.0

    object_weight: 100
    object_tolerance: 0.01

    grasp_translation_offset: [ 0.0, 0.0, 0.0 ]
    grasp_orientation_offset: [ 0, -0.7071068, 0, 0.7071068 ]    # -90deg around y

    # Safer joint limits (+10 deg safety margin ~0.17)
    upper_joint_limits: [ 2.0, 2.0, 6.28, 2.72, 1.59, 2.72, -0.1, 2.72, 3.58, 2.72 ]
    lower_joint_limits: [ -2.0, -2.0, -6.28, -2.72, -1.59, -2.72, -2.90, -2.55, 0.152, -2.72 ]

    joint_limit_cost: 1000 #1000
    joint_limit_slope: 100 #100

    max_power: 0.0
    power_weight: 0.0
  
    handle_frame: handle_link
    tracked_frame: panda_grasp_finger_edge
    #tracked_frame: panda_grasp
    arm_base_frame: panda_link0

    collision_link_0: panda_link1
    collision_link_1: panda_link7

    collision_threshold: 0.35
    collision_weight: 1000


  controller_gains:
    base_gains:
      kp: [ 0.0, 0.0, 0.0 ]
      kd: [ 0.0, 0.0, 0.0 ]
      ki: [ 0.0, 0.0, 0.0 ]
      i_max: [ 0.0, 0.0, 0.0 ]
    arm_gains:
      kp: [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ]
      kd: [ 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0 ]
      ki: [ 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0 ]
      i_max: [ 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 ]
    gripper_gains:
      kp: [ 100.0, 100.0 ]
      kd: [ 50.0, 50.0 ]
      ki: [ 0.0, 0.0 ]
      i_max: [ 0.0, 0.0 ]

  object_reference_position: 1.57

  # Parameters of the dynamics internal to the controller
  dynamics:
    dt: 0.015
    object_handle_link: handle_link
    object_handle_joint: handle_joint
    articulation_joint: articulation_joint

    initial_state: [ 0.0, 0.0, 0.0, 0.0, -0.52, 0.0, -1.785, 0.0, 1.10, 0.69, 0.04, 0.04,    # robot position
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,             # robot velocity
                     0.0,                                                                    # object joint position
                     0.0,                                                                    # object joint velocity
                     0,                                                                      # contact flag
                     0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ]            # external torques

    gains:
      base_gains:
        kp: [ 0.0, 0.0, 0.0 ]
        kd: [ 1000.0, 1000.0, 1000.0 ]
        ki: [ 0.0, 0.0, 0.0 ]
        i_max: [ 0.0, 0.0, 0.0 ]
      arm_gains:
        kp: [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ]
        kd: [ 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0 ]
        ki: [ 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1 ]
        i_max: [ 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 ]
      gripper_gains:
        kp: [ 100.0, 100.0 ]
        kd: [ 50.0, 50.0 ]
        ki: [ 0.0, 0.0 ]
        i_max: [ 0.0, 0.0 ]